rfd/0096-helm-chart-revamp.md

```diff
@@ -0,0 +1,400 @@
+but allow users to pass arbitrary additional configuration or perform specific
+overrides.  This way, users would not have to leave the happy path if they need
+to set one specific value.  Manually implementing all configuration knobs in
+Helm adds no value and brings confusion as some values are not supported or not
+named the same way than the `teleport.yaml` field they set.
+
+By leveraging Helm's templating functions `toYaml`, `fromYaml`, and sprig's
+`mustMergeOverwrite` the charts would merge their automatically-generated
+`teleport.yaml` with user-provided `teleport.yaml`.
+
+A user deploying the `auth_service` chart in `standalone` mode and wanting to
+set key exchange algorithms, remove extra log fields, and override kube
+cluster's name would use the values:
+
+```yaml
+clusterName: my-cluster
+chartMode: standalone
+auth:
+  teleportConfig:
+    teleport:
+      kex_algos:
+        - ecdh-sha2-nistp256
+        - ecdh-sha2-nistp384
+        - ecdh-sha2-nistp521
+      log:
+        format:
+          extra_fields: ~
+    kubernetes_service:
+      kube_cluster_name: my-override
+```
+
+The generated chart configuration for the `standalone` mode is
+```yaml
+teleport:
+  log:
+    severity: INFO
+    output: stderr
+    format:
+      output: text
+      extra_fields: ["timestamp","level","component","caller"]
+auth_service:
+  enabled: true
+  cluster_name: my-cluster
+  authentication:
+    type: "local"
+    local_auth: true
+    second_factor: "otp"
+kubernetes_service:
+  enabled: true
+  listen_addr: 0.0.0.0:3027
+  kube_cluster_name: my-cluster
+proxy_service:
+  enabled: false
+ssh_service:
+  enabled: false
+```
+
+Once merged with the custom user configuration, the resulting configuration is
+```yaml
+auth_service:
+  authentication:
+    local_auth: true
+    second_factor: otp
+    type: local
+  cluster_name: my-cluster
+  enabled: true
+kubernetes_service:
+  enabled: true
+  kube_cluster_name: my-override
+  listen_addr: 0.0.0.0:3027
+proxy_service:
+  enabled: false
+ssh_service:
+  enabled: false
+teleport:
+  kex_algos:
+    - ecdh-sha2-nistp256
+    - ecdh-sha2-nistp384
+    - ecdh-sha2-nistp521
+  log:
+    format:
+      extra_fields: null
+      output: text
+    output: stderr
+    severity: INFO
+```
+
+The proof of concept code [can be found
+here](https://github.com/hugoShaka/teleport-helm-config-poc).
+
+The main drawback of this approach is that comments and value ordering are lost
+during the round-trip.  This approach could be extended to support multiple
+configuration syntax, following a breaking change for example.
+
+`custom` should be removed in favor of a new `scratch` mode. Compared to
+the previous Helm chart, users would not provide an external ConfigMap
+but pass the custom configuration through the values. This is a breaking
+change for them, but by the nature of the auth/proxy split it is not possible
+to be backward compatible with `custom` mode.
```

Comment:
Are we intending to add a migration guide from a `custom` config to the new format in the docs? I know a lot of people are already using `custom` and so it'd be nice to guide them through the process to ease the pain of a breaking change.

Reply:
Like this guide? https://github.com/gravitational/teleport/pull/19881/files#diff-c7dd7c2ab1f43e582d7662823d997dc4317c42d334fb179a4b3a5cd4dbc911dd